src/client/CodeEditor/widgets.ts

```diff
@@ -383,3 +388,100 @@ export const highlightWidgets = ViewPlugin.fromClass(
     },
   },
 );
+
+export const rotationIndicator = ViewPlugin.fromClass(
+  class {
+    view: EditorView;
+    textPosition?: number;
+    rotation: number;
+    constructor(view) {
+      this.view = view;
+      this.textPosition = null;
+      this.rotation = 0;
+
+      document.addEventListener(
+        'updateRotateDrag',
+        (e: CustomEvent) => {
+          this.textPosition = this.view.posAtCoords(
+            rotationOrigin,
+            false,
+          );
+
+          this.rotation = e.detail;
+        },
+      );
+    }
+  },
+  {
+    decorations: (v) => {
+      if (rotationOrigin === null) {
+        return Decoration.none;
+      }
+      return Decoration.set([
+        {
+          from: v.textPosition,
+          to: v.textPosition,
+          value: Decoration.widget({
+            side: -1,
+            widget: new RotationCircle(v.rotation),
+          }),
+        },
+      ]);
+    },
+  },
+);
+
+class RotationCircle extends WidgetType {
+  angle: number;
+  constructor(angle: number) {
+    super();
+    this.angle = angle;
+  }
+
+  toDOM(view: EditorView): HTMLElement {
+    const parent = document.createElement('div');
```

Comment:
It looks like this creates a new DOM element each time you move the mouse. It's not a big deal as performance seems fine, but something I wonder if we could optimize by re-using the existing DOM element and just changing the indicator line.